Improve search history attaching change listener

demetres12 opened the pull request:
Proposes solution for #9685 

Initializing the draft pull request regarding the improvement of search history.

Specifically,  redundant recording of search queries, issued via:
- attachment of a change listener on focusedProperty of search bar textfield, and by
- using unidirectional binding between searcField  and a BooleanPorperty object

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

```[tasklist]
### Compulsory checks
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.
```

Siedlerchr commented on src/main/java/org/jabref/gui/StateManager.java:
```diff
@@ -216,8 +217,10 @@ public List<String> collectAllDatabasePaths() {
     public void addSearchHistory(String search) {
-        searchHistory.remove(search);
-        searchHistory.add(search);
+        if (!GlobalSearchBar.searchBoxFocused.get()) {
```
I would not add this here. The state manager should not have a dependency on the search bar. This is also not relevant here.

demetres12 replied:
At first I just attached the change listener inside GlobalSearchBar, but I was concerned about the case we call the addSearchHistory from elsewhere, and passed the bound object to StateManager.

So, is there any chance the method to be called from another class to record a query?

Siedlerchr replied:
No, why should we? Tip: Every IDE has a thing called "Call Hierarchy" to show where the method is called. Use this to check where the method is called

demetres12 replied:
I placed the change listener under constructor. I'm not sure what to do inside the if statement:
```
searchBoxFocused.bind(searchField.focusedProperty());
searchBoxFocused.addListener((observable, oldPropertyValue, newPropertyValue) -> {
    if (!newPropertyValue) {
        // do sth
    }
});
```
About performSearch method:
```
if (!(searchBoxFocused.get() || searchField.textProperty().get().isEmpty())) {
    this.stateManager.addSearchHistory(searchField.textProperty().get());
}

Siedlerchr replied:
The focus listener gives you two values:
old and new
aka: the previous value and the new value:
So: When the focus is lost: check if old = true && new = false  
=> The user has left the search box
=> Now we call the addSearchHistory with the value of the     this.stateManager.addSearchHistory(searchField.textProperty().get());

demetres12 replied:
Yes I get it. 

The change listener is placed under constructor. The method addSearchHistory, in contrast, is called inside performSearch(), so that's the reason I use binding to attach the value of the focus property to another BooleanProperty object's value.

As you can see from the two blocks of code (my previous comment), both if statements do the same job. Checking whether the focus is lost or not. So that's why I'm wondering what to do in the first if statement.

Siedlerchr replied:
In this case you neither need the binding nor the listener.

My approach would have been
```
searchBoxFocused.addListener((obs, oldValue, newValue) -> {
    if (oldValue  && !new  &&  && ! searchField.textProperty().get().isEmpty())) {
    this.stateManager.addSearchHistory(searchField.textProperty().get());
    }
});
```
remove the existing one in peformSearch

demetres12 replied:
> In this case you neither need the binding nor the listener.

Nice, it seems to works correctly. I will write some javadoc (regarding
the classes I touched) and unit tests.

> remove the existing one in performSearch

Actually, removing `addSearchHistory` method call from the`performSearch` 
prevents the unnecessary  recording of history every the time the `searchTask` timer
exceeds the  (`SEARCH_DELAY = 400`)  and `performSearch` is called.

Siedlerchr commented on src/main/java/org/jabref/gui/search/GlobalSearchBar.java:
```diff
@@ -107,6 +107,7 @@ public class GlobalSearchBar extends HBox {
     private final DialogService dialogService;
 
     private final BooleanProperty globalSearchActive = new SimpleBooleanProperty(false);
+    public static final BooleanProperty searchBoxFocused = new SimpleBooleanProperty(false);
```
See as above, it's not necessary to expose this

Siedlerchr commented on src/main/java/org/jabref/gui/search/GlobalSearchBar.java:
```diff
@@ -307,7 +308,12 @@ public void performSearch() {
             return;
         }
-        this.stateManager.addSearchHistory(searchField.textProperty().get());
+        searchBoxFocused.bind(searchField.focusedProperty());
```
this whole block needs to be outside the performSearch.
Currently you add a new listener for every search. But we need to have only one listener. See e.g. line 212

koppor commented:
You got partial feedback at https://github.com/JabRef/jabref/issues/9685#issuecomment-1519713484.

To judge the question, I need to check the current state. Maybe, you can summarize it? Not sure, when I have time to checkout your PR and check it. Not sure about the others though.

demetres12 commented:
Briefly, regarding the current state, the bug is fixed and there are no redundant recordings.
The bug as introduced in issue's description, was the following (at my understanding). The addSearchHistory method call, as embedded in performSearch, was called every time the search box was idle for more than 400ms. That caused issues (e.g. prefixes were stored in case we slowtype). So, Houssem suggested two solutions regarding search history improvement.
### First solution 
**Attach a change listener** to focus property of search box node, so only when it gets inactive / lost (i.e. user completes search) then we record the query. Also removed the addSearchHistory method call from performSearch and placed it under constructor. See pr's description for more details
### Second solution
Each time a query is recorded, we loop the search history observable list and: 
- Check **StringSimilarity** between the recently added and the previous ones, by calculating the edit distance, that needs to be within the preferred threshold.
- We also check if the **time period** between the two recordings is short enough.

In case both of them return the desired results, we merge the two queries
### What I implemented?
**I followed the first solution**.  
I will also upload some screenshots soon enough.

Siedlerchr commented:
Totally fine if you don't need the similarity. I'm happy with the solution.
Try to make a test for it, if it's too complicated (gui tests are not easy, a manual test is also fine)

koppor commented on src/main/java/org/jabref/gui/search/GlobalSearchBar.java:
```diff
@@ -211,6 +211,20 @@ public GlobalSearchBar(JabRefFrame frame, StateManager stateManager, Preferences
         this.stateManager.activeSearchQueryProperty().addListener((obs, oldvalue, newValue) -> newValue.ifPresent(this::updateSearchResultsForQuery));
         this.stateManager.activeDatabaseProperty().addListener((obs, oldValue, newValue) -> stateManager.activeSearchQueryProperty().get().ifPresent(this::updateSearchResultsForQuery));
+        /*
+         * Attach a change listener on the focus property of searchField node.
```
This is clear from the code, this line can be removed.

koppor commented on src/main/java/org/jabref/gui/search/GlobalSearchBar.java:
```diff
@@ -211,6 +211,20 @@ public GlobalSearchBar(JabRefFrame frame, StateManager stateManager, Preferences
         this.stateManager.activeDatabaseProperty().addListener((obs, oldValue, newValue) -> stateManager.activeSearchQueryProperty().get().ifPresent(this::updateSearchResultsForQuery));
+        /*
+         * Attach a change listener on the focus property of searchField node.
+         * Every time the listener tracks a change on focus property value,
+         * (i.e. from active (user types query) to inactive / lost (user selects
+         * entry / completes search)), ensures the focus is lost
+         * (i.e. newValue : false), checks whether the search query is
+         * empty or not and finally records it into the search history
+         * through addSearchHistory method call.
```
I would use line breaks to guide the reader. Moreover, the sentence is really too long to understand.

The text should not appear as generated one-by-one by reading the code.

Shorter version:

```suggestion
         * The listener tracks a change on the focus property value.
         * This happens, from active (user types a query) to inactive / focus lost (e.g., user selects an entry or triggers the search).
         * The search history should only be filled, if focus is lost.
```

koppor commented on src/main/java/org/jabref/gui/search/GlobalSearchBar.java:
```diff
@@ -211,6 +211,20 @@ public GlobalSearchBar(JabRefFrame frame, StateManager stateManager, Preferences
+         * through addSearchHistory method call.
+         */
+        searchField.focusedProperty().addListener((obs, oldValue, newValue) -> {
+            if (oldValue && !(newValue || searchField.getText().isEmpty())) {
```
Here you can add the detailed comment:

// Focus lost can derived by checking that there is no newValue (or the text is empty)

koppor reviewed (changes requested):
"Code" comments attached.

koppor commented:
I tested it, but could not see any search history. What do I need to do to test this feature?

![image](https://user-images.githubusercontent.com/1366654/235269396-b042a06d-07dd-4b8b-9fe6-8daf54ecbcd6.png)

Is there any screenshot?

demetres12 commented:
@koppor I have just uploaded screenshots.
@Siedlerchr Will also add GUI tests so please don't merge.

koppor commented:
Thank you for the video! -- I was expecting that there was a dropdown shown. Need to search the PR and check for the decision.

koppor reviewed (commented):
Improves the situation, so I would suggest to add a CHANGELOG entry and then we go for a merge. Tests can be done later.

Siedlerchr commented on CHANGELOG.md:
```diff
@@ -66,6 +66,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where searching for unlinked files would include the current library's .bib file [#9735](https://github.com/JabRef/jabref/issues/9735)
 - We fixed an issue where it was no longer possible to connect to a shared mysql database due to an exception [#9761](https://github.com/JabRef/jabref/issues/9761)
 - We fixed the citation key generation for (`[authors]`, `[authshort]`, `[authorsAlpha]`, `authIniN`, `authEtAl`, `auth.etal`)[https://docs.jabref.org/setup/citationkeypatterns#special-field-markers] to handle `and others` properly. [koppor#626](https://github.com/koppor/jabref/issues/626)
+- We fixed an issue regarding recording redundant prefixes in search history, by attaching a change listener on the focus property of search box. [#9685](https://github.com/JabRef/jabref/issues/9685)
```
```suggestion
- We fixed an issue regarding recording redundant prefixes in search history [#9685](https://github.com/JabRef/jabref/issues/9685)
```

The changelog is intended for the "normal end user"  so no  technical details necessary

Siedlerchr reviewed (dismissed):
From my side lgtm so far already

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @param robot Reference to the FxRobot class, a TestFX utility class
+     */
+
```
Please no empty line between JavaDoc and method where the javadoc belongs.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+        var searchFieldRoboto = robot.clickOn(searchField);
+        for (char c : searchQuery.toCharArray()) {
+            searchFieldRoboto.write(String.valueOf(c));
+            try {
+                Thread.sleep(401);
+                Assertions.assertTrue(stateManager.getWholeSearchHistory().isEmpty());
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
```
No need for try-catch in Test cases. Just add ` throws Exception` to the method signature.

I don't think, that `InterruptedException` should be ignored and then the test continue.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @param robot Reference to the FxRobot class, a TestFx utility class
+     */
+
```
Remove empty line

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+ * that neither redundant data are recorded to the search history nor
+ * empty search queries are stored.
+ *
+ * @author dkokkotas
+ * @version 1.0
+ * @since May 2023
```
Remove these - we don't use these JavaDoc annotations.

Moreover `@since` indicaes API availability. And there is no user-visible API of tests.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+ * @version 1.0
+ * @since May 2023
+ */
+
```
Remove empty line

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+
+/**
+ * The GlobalSearchBarTest class currently serves the GUI testing regarding
+ * search history recording improvement. Designed in TestFX testing framework
```
Remove "improvement". A test tests current program - not talking about past and future.

I think, the comment on FxRobot is not necessary, because whole JabRef UI test relies on FxRobot. 

You can write documentation to https://github.com/JabRef/jabref/blob/main/docs/code-howtos/testing.md if you want.

In case you want to compare FxRobot with other possibilities, you can write an architecture-decision-record (https://github.com/JabRef/jabref/tree/main/docs/decisions) if you want.

koppor replied:
I think, the whole JavaDoc comment can be deleted. The content can be derived from a) the class name and b) from general knowledge of JabRef testing.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,156 @@
+    private GlobalSearchBar searchBar;
+    private StateManager stateManager;
+
+    /**
+     * The onStart method, accepts the stage of the application and sets
+     * up the basic components of the UI (i.e. scene, layout), that will
+     * be used for our tests. Also interacts with GlobalSearchBar and
+     * StateManager classes, in order to set up the search box.
+     *
+     * @param stage Reference to the Stage class, part of the JavaFX UI toolkit
+     */
```
May I ask if this comment is really necessary? Doesn't one see it directly from the code?

Reason of minimizing comments: If the code is updated, sometimes the JavaDoc is not updated, and then the whole thing diverges, which leads to higher effort for maintenance.

demetres12 replied:
> May I ask if this comment is really necessary? Doesn't one see it directly from the code?

I just considered an explanatory documentation both on class and it's methods to be appropriate. However, if a detailed approach is redundant and causes higher effort for maintenance, it must be removed or shortened.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,128 @@
+            Assertions.assertTrue(stateManager.getWholeSearchHistory().isEmpty());
+        }
+
+        /*
+         * Set the focus to another node to trigger the listener and finally
+         * record the query.
+         */
```
Please try to avoid these kind of multine line comments.

You can easily use single line comments

```suggestion
        // Set the focus to another node to trigger the listener and finally record the query.
```

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,128 @@
+        List<String> lastSearchHistory = stateManager.getWholeSearchHistory();
+        List<String> expected = List.of("Smith");
+
+        Assertions.assertEquals(expected, lastSearchHistory);
```
See the other tests - please do a static import of `assertEquals`

koppor replied:
This test fails here:

```
org.opentest4j.AssertionFailedError: expected: java.util.ImmutableCollections$List12@3bfad369<[Smith]> but was: com.sun.javafx.collections.ObservableListWrapper@25e85447<[Smith]>
Expected :[Smith]
Actual   :[Smith]
```

I assume, you have to do:

    assertEquals(FXCollections.observableArrayList("Smith"), lastSearchHistory);

(No need to declare the expected variable - it is less than 20 characters code in your case. And in my suggestion, it it sstill readable).

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,128 @@
+        Assertions.assertEquals(expected, lastSearchHistory);
+    }
+
+    /**
+     * The avoidRecordingEmptyQuery method, accepts a robot reference to the
+     * FxRobot class and verifies that no empty search queries are recorded.
+     *
+     * @param robot Reference to the FxRobot class, a TestFX utility class
+     */
+    @Test
+    void avoidRecordingEmptyQuery(FxRobot robot) {
```
The test method title should be clear. The JavaDoc can be removed - the parameter `robot` is the same for all test classes - and even a generael concept. Maybe add that to our test documentation?

```suggestion
    @Test
    void emptyQueryIsNotRecorded(FxRobot robot) {
```

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,128 @@
+        stage.show();
+    }
+
+    /**
+     * The recordingSearchQueriesOnFocusLost method, accepts a robot reference
+     * to the FxRobot class and verifies that unnecessary prefixes of the
+     * search query are not recorded. Essentially, the query is recorded
+     * only when the searchField node loses focus.
+     *
+     * @param robot Reference to the FxRobot class, a TestFx utility class
+     */
+    @Test
+    void recordingSearchQueriesOnFocusLost(FxRobot robot) throws InterruptedException {
```
(Explanation: see below)

```suggestion
    @Test
    void recordingSearchQueriesOnFocusLostOnly(FxRobot robot) throws InterruptedException {
```

koppor reviewed (changes requested):
Some smaller comments - especially because of the failing test.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,110 @@
+        DefaultTaskExecutor.runInJavaFXThread(() -> hBox.requestFocus());
+        List<String> lastSearchHistory = stateManager.getWholeSearchHistory();
+
+        assertEquals(FXCollections.observableArrayList("Smith"), lastSearchHistory);
```
Test fails:

    expected: com.sun.javafx.collections.ObservableListWrapper@34f97e7e<[Smith]> but was: com.sun.javafx.collections.ObservableListWrapper@5af934eb<[Smith]>

When playing around, I even did not get any lastSearchHistory

My current code:

```java
        List<String> lastSearchHistory = stateManager.getWholeSearchHistory().stream().toList();

        assertEquals(List.of("Smith"), lastSearchHistory);
```

Did the test ever succeed on your side?

demetres12 replied:
Both tests are successfully passed on my side

demetres12 replied:
A reasonable approach could be that  the `assertEquals` method handles the two arguments as different types of objects, something that  could cause the test to fail. However, `stateManager.getWholeSearchHistory()` also returns an observable list.

demetres12 replied:
@Siedlerchr do you face same issues?

Siedlerchr replied:
Last time I tried the test it was working for me as well.
Can check later this evening again

koppor commented:
Regarding the tests. A screenshot of current output. One needs to scroll down to see the failing test (scrollbar marked in the screenshot)

![image](https://user-images.githubusercontent.com/1366654/235848336-6c9f088c-6b85-4e30-aa96-17911f31828c.png)

When clicking on "Details" it shows:

![image](https://user-images.githubusercontent.com/1366654/235848388-29f076ca-029c-4895-95b5-4bbc96cde2e7.png)

Thus, my proposal to use `FXCollections.observableArrayList("Smith")` did not work.

@Siedlerchr Note that both types are `com.sun.javafx.collections.ObservableListWrapper` already. Seems they are not comparable. Therefore, I went to https://stackoverflow.com/q/39872697/873282, but our objects do not have `getItems()`. Therefore, I proposed to use 

    List<String> lastSearchHistory = stateManager.getWholeSearchHistory().stream().toList();

     assertEquals(List.of("Smith"), lastSearchHistory);

I tried only locally, but not on GitHub.

koppor commented:
The source of `ObservableListWrapper` also contains no comparable: https://github.com/openjdk/jfx/blob/master/modules/javafx.base/src/main/java/com/sun/javafx/collections/ObservableListWrapper.java. Do we need to contribute to JavaFX code?

koppor commented:
Good to hear that it works on your macOS systems (you and @demetres12 use macOS). -- It does not work on my Windows and it also does not work on GitHub.

For GitHub, I already showed the links to the failing tests.

It is totally unclear to me why this happens.

@calixtus Can you run [GlobalSearchBarTest.java](https://github.com/JabRef/jabref/pull/9794/files#diff-299e751d310989961836a0bda41eddaaf39798e9827684a357c475e881262a78) on your Linux (and maybe Windows) machine?

On the one hand, it would be OK for me to disable the tests and to reserve it as "future work" to fix them. On the other hand, working with FXCollections in tests is very fundamental. Thus, we should devote some time to get this working on GitHub (Linux!) and on Windows machines.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -88,8 +89,8 @@ void recordingSearchQueriesOnFocusLostOnly(FxRobot robot) throws InterruptedExce
-        List<String> lastSearchHistory = stateManager.getWholeSearchHistory();
-        List<String> expected = List.of("Smith");
+        ObservableList<String> lastSearchHistory = stateManager.getWholeSearchHistory();
+        ObservableList<String> expected = FXCollections.observableArrayList("Smith");
```
The type is still `ObservableListWrapper`.

org.opentest4j.AssertionFailedError: expected: com.sun.javafx.collections.ObservableListWrapper@105ad36f<[Smith]> but was: com.sun.javafx.collections.ObservableListWrapper@dfc809b<[Smith]>.

I still think, ´stateManager.getWholeSearchHistory()` should be converted to a `List`. (For instance, using `.stream().toList()`.

demetres12 replied:
Using `stream().toList()` causes issues and doesn't return the search history elements as expected.

Siedlerchr commented:
change the Focus method to      DefaultTaskExecutor.runAndWaitInJavaFXThread(() -> hBox.requestFocus());
 

that seemed to do the trick. Wating for the FX thread to execute

Siedlerchr commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -88,8 +89,8 @@ void recordingSearchQueriesOnFocusLostOnly(FxRobot robot) throws InterruptedExce
         // Set the focus to another node to trigger the listener and finally record the query.
         DefaultTaskExecutor.runInJavaFXThread(() -> hBox.requestFocus());
-        List<String> lastSearchHistory = stateManager.getWholeSearchHistory();
-        List<String> expected = List.of("Smith");
```
You can keep the List interface. The test failure was misleading. It turned out that it was empty:
you can add an assertFalse(lastSearchHistory.isEmpty())

demetres12 replied:
Works fine!

Siedlerchr commented:
Yeah seems like it worked now 👍  💯

Siedlerchr commented:
Basically, this is a feature of junit5. Back in JUnit 4 they had to be public. 

> It is generally recommended to omit the public modifier for test classes, test methods, and lifecycle methods unless there is a technical reason for doing so – for example, when a test class is extended by a test class in another package. Another technical reason for making classes and methods public is to simplify testing on the module path when using the Java Module System.

https://junit.org/junit5/docs/snapshot/user-guide/

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,112 @@
+        DefaultTaskExecutor.runAndWaitInJavaFXThread(() -> hBox.requestFocus());
+        List<String> lastSearchHistory = stateManager.getWholeSearchHistory().stream().toList();
+
+        assertTrue(lastSearchHistory.isEmpty());
```
As test pattern, one should compare lists. This helps if the test fails. Then, in the test diff, one can directly see the output.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,112 @@
+        DefaultTaskExecutor.runAndWaitInJavaFXThread(() -> hBox.requestFocus());
+        List<String> lastSearchHistory = stateManager.getWholeSearchHistory().stream().toList();
+
+        assertFalse(lastSearchHistory.isEmpty());
```
This is not necessary, the next line already tests for emptyness.

Siedlerchr replied:
No, this check was necessary to check for this weird behavior in the assertEquals and to ensure that the focus lost was working

koppor replied:
Thus, it was just a left-over 😅. Now, everything works. The execution time of this assertation statement should not have an impact on the equals check. I noted that the same list is checked here and in the next line.

koppor commented on src/test/java/org/jabref/gui/search/GlobalSearchBarTest.java:
```diff
@@ -0,0 +1,112 @@
+        List<String> lastSearchHistory = stateManager.getWholeSearchHistory().stream().toList();
+
+        assertFalse(lastSearchHistory.isEmpty());
+        assertEquals(FXCollections.observableArrayList("Smith"), lastSearchHistory);
```
Here, `List.of("Smith")` no works, no need to use `FXCollections`.

koppor reviewed (changes requested):
Thank you for the updates. I have three nitpicks on the testing. I commented - and also updated the code for myself to keep things going. If the tests are green, I will merge.

demetres12 commented:
@koppor @Siedlerchr
Your expertise and enthusiasm were both appreciated during a time that can be challenging for my concerns!!!
Heart thanks and gratefulness
